File: src/previous/day72/Map.ts
type Item<K, V> = { key: K; value: V }
type Bucket<K, V> = Array<Item<K, V>>

export default class Map<T extends string | number, V> {
  private length: number
  private capacity: number
  private store: Array<Bucket<T, V>>
  constructor(capacity: number = 3) {
    this.length = 0
    this.capacity = capacity
    this.store = Array.from({ length: capacity }, () => [])
  }

  get(key: T): V | undefined {
    const hash = this.hash(key)
    const bucket = this.store[hash]

    for (let i = 0; i < bucket.length; i++) {
      if (key !== bucket[i].key) continue
      return bucket[i].value
    }

    return undefined
  }

  set(key: T, value: V): void {
    const hash = this.hash(key)
    const bucket = this.store[hash]

    for (let i = 0; i < bucket.length; i++) {
      if (key !== bucket[i].key) continue
      this.resize()
      bucket[i].value = value
      return
    }

    this.length++
    bucket.push({ key, value })
  }
  delete(key: T): V | undefined {
    const hash = this.hash(key)
    const bucket = this.store[hash]

    for (let i = 0; i < bucket.length; i++) {
      if (key !== bucket[i].key) continue

      const out = bucket[i].value
      this.length--
      bucket[i] = bucket[bucket.length - 1]
      bucket.pop()
      return out
    }

    return undefined
  }
  size(): number {
    return this.length
  }

  private hash(key: T): number {
    return (
      key
        .toString()
        .split("")
        .reduce((acc, char) => acc + char.charCodeAt(0) + 29, 3) % this.capacity
    )
  }

  private resize(): void {
    if (this.length <= this.capacity) return

    const data = this.store
    this.capacity *= 2
    this.store = Array.from({ length: this.capacity }, () => [])

    for (let i = 0; i < data.length; i++) {
      const bucket = data[i]
      for (let k = 0; k < bucket.length; k++) {
        const hash = this.hash(bucket[k].key)
        this.store[hash].push(bucket[k])
      }
    }

    console.log("Map resized", this.store)
  }
}

File: go/day02/Map/Map.go
package basic_map

import (
	"fmt"
)

type Map[K, V comparable] struct {
	length   int
	capacity int
	store    []Bucket[K, V]
}

type Item[K, V comparable] struct {
	key   K
	value V
}

type Bucket[K, V comparable] []Item[K, V]

func MakeMap[K, V comparable]() *Map[K, V] {
	store := make([]Bucket[K, V], 3)

	return &Map[K, V]{
		length:   0,
		capacity: 3,
		store:    store,
	}
}

func (m *Map[K, V]) Get(key K) V {
	hash := m.hash(key)
	bucket := m.store[hash]

	return bucket[0].value
}

func (m *Map[K, V]) Set(key K, value V) {
	hash := m.hash(key)
	bucket := m.store[hash]

	for i, item := range bucket {
		if item.key == key {
			bucket[i].value = value
			return
		}
	}

	m.store[hash] = append(bucket, Item[K, V]{key, value})

	m.length++
	m.resize()
}

func (m *Map[K, V]) Delete(key K) V {
	if m.length == 0 {
		var z V
		return z
	}

	hash := m.hash(key)
	bucket := m.store[hash]

	for _, item := range bucket {
		if item.key == key {
			m.length--
			m.store[hash] = m.store[hash][:0]
			return item.value
		}
	}

	var z V
	return z
}

func (m *Map[K, V]) Size() int {
	return m.length
}

func (m *Map[K, V]) resize() {
	if m.length <= m.capacity {
		return
	}

	store := m.store
	m.capacity *= 2
	m.store = make([]Bucket[K, V], m.capacity)

	for _, bucket := range store {
		for _, item := range bucket {
			hash := m.hash(item.key)
			m.store[hash] = append(m.store[hash], item)

		}
	}
}

func (m *Map[K, V]) hash(key K) int {
	keyStr := fmt.Sprintf("%v", key)
	acc := 37

	for _, char := range keyStr {
		acc += int(char) + 7
	}

	return acc % m.capacity
}
